File: backend/app/schemas/viral_analyzer.py
"""
爆款模仿相关 Schema
分析爆款文章风格并生成类似内容
"""
from typing import List, Optional
from pydantic import BaseModel, Field
from enum import Enum


class ContentCategory(str, Enum):
    """内容类别"""
    EMOTION = "emotion"           # 情感共鸣
    KNOWLEDGE = "knowledge"       # 知识干货
    STORY = "story"               # 故事叙事
    NEWS = "news"                 # 新闻热点
    TUTORIAL = "tutorial"         # 教程攻略
    REVIEW = "review"             # 测评种草
    OPINION = "opinion"           # 观点输出


class ViralElement(BaseModel):
    """爆款元素"""
    name: str = Field(..., description="元素名称")
    description: str = Field(..., description="元素说明")
    score: int = Field(..., description="该元素的运用得分 (0-100)")
    examples: List[str] = Field(default_factory=list, description="文中示例")


class StructureAnalysis(BaseModel):
    """结构分析"""
    sections: List[str] = Field(..., description="段落结构")
    opening_hook: str = Field(..., description="开头钩子")
    closing_cta: str = Field(..., description="结尾行动召唤")
    transition_style: str = Field(..., description="过渡风格")


class AnalyzeRequest(BaseModel):
    """爆款分析请求"""
    content: str = Field(..., description="要分析的爆款内容", min_length=50, max_length=20000)
    title: Optional[str] = Field(None, description="文章标题")
    platform: Optional[str] = Field(None, description="来源平台")


class AnalyzeResponse(BaseModel):
    """爆款分析响应"""
    title: str = Field(..., description="文章标题")
    category: ContentCategory = Field(..., description="内容类别")
    viral_score: int = Field(..., description="爆款指数 (0-100)")
    creation_id: Optional[int] = Field(None, description="创作记录ID（落库后返回）")
    
    # 风格分析
    tone: str = Field(..., description="语气风格")
    target_audience: str = Field(..., description="目标受众")
    emotional_triggers: List[str] = Field(..., description="情感触发点")
    
    # 爆款元素
    viral_elements: List[ViralElement] = Field(..., description="爆款元素分析")
    
    # 结构分析
    structure: StructureAnalysis = Field(..., description="结构分析")
    
    # 写作技巧
    writing_techniques: List[str] = Field(..., description="写作技巧")
    
    # 关键词
    keywords: List[str] = Field(..., description="核心关键词")
    
    # 改进建议
    improvement_suggestions: List[str] = Field(default_factory=list, description="可改进的点")


class ImitateRequest(BaseModel):
    """爆款模仿请求"""
    reference_content: str = Field(..., description="参考的爆款内容", min_length=50, max_length=20000)
    reference_title: Optional[str] = Field(None, description="参考文章标题")
    new_topic: str = Field(..., description="新的主题/话题", min_length=2, max_length=200)
    platform: Optional[str] = Field(None, description="目标平台")
    style_strength: int = Field(80, description="风格模仿强度 (0-100)", ge=0, le=100)
    keep_structure: bool = Field(True, description="是否保持相同结构")
    additional_requirements: Optional[str] = Field(None, description="额外要求")


class ImitateResponse(BaseModel):
    """爆款模仿响应"""
    title: str = Field(..., description="生成的标题")
    content: str = Field(..., description="生成的内容")
    
    # 模仿说明
    imitation_notes: List[str] = Field(..., description="模仿说明")
    elements_applied: List[str] = Field(..., description="应用的爆款元素")
    
    # 统计信息
    word_count: int = Field(..., description="字数")
    estimated_viral_score: int = Field(..., description="预估爆款指数")
    
    # 创作记录ID
    creation_id: Optional[int] = Field(None, description="创作记录ID")


class StyleTemplateRequest(BaseModel):
    """风格模板提取请求"""
    contents: List[str] = Field(..., description="多篇同风格的内容", min_items=2, max_items=5)
    template_name: str = Field(..., description="模板名称")


class StyleTemplate(BaseModel):
    """风格模板"""
    id: Optional[int] = Field(None, description="模板ID")
    name: str = Field(..., description="模板名称")
    description: str = Field(..., description="模板描述")
    
    # 风格特征
    tone: str = Field(..., description="语气风格")
    sentence_style: str = Field(..., description="句式特点")
    vocabulary_style: str = Field(..., description="用词风格")
    structure_pattern: str = Field(..., description="结构模式")
    
    # 爆款元素配比
    element_weights: dict = Field(..., description="元素权重配置")
    
    # 示例句式
    example_sentences: List[str] = Field(..., description="示例句式")


class QuickImitateRequest(BaseModel):
    """快速模仿请求（基于模板）"""
    template_id: int = Field(..., description="风格模板ID")
    topic: str = Field(..., description="新主题")
    word_count: int = Field(800, description="目标字数", ge=100, le=5000)

File: frontend/src/api/viralAnalyzer.ts
/**
 * 爆款模仿 API
 * 分析爆款文章风格并生成类似内容
 */
import request from './request'

// 内容类别枚举
export enum ContentCategory {
  EMOTION = 'emotion',           // 情感共鸣
  KNOWLEDGE = 'knowledge',       // 知识干货
  STORY = 'story',               // 故事叙事
  NEWS = 'news',                 // 新闻热点
  TUTORIAL = 'tutorial',         // 教程攻略
  REVIEW = 'review',             // 测评种草
  OPINION = 'opinion',           // 观点输出
}

// 爆款元素
export interface ViralElement {
  name: string                   // 元素名称
  description: string            // 元素说明
  score: number                  // 该元素的运用得分 (0-100)
  examples: string[]             // 文中示例
}

// 结构分析
export interface StructureAnalysis {
  sections: string[]             // 段落结构
  opening_hook: string           // 开头钩子
  closing_cta: string            // 结尾行动召唤
  transition_style: string       // 过渡风格
}

// 爆款分析请求
export interface AnalyzeRequest {
  content: string                // 要分析的爆款内容
  title?: string                 // 文章标题
  platform?: string              // 来源平台
}

// 爆款分析响应
export interface AnalyzeResponse {
  title: string                  // 文章标题
  category: ContentCategory      // 内容类别
  viral_score: number            // 爆款指数 (0-100)
  
  // 风格分析
  tone: string                   // 语气风格
  target_audience: string        // 目标受众
  emotional_triggers: string[]   // 情感触发点
  
  // 爆款元素
  viral_elements: ViralElement[] // 爆款元素分析
  
  // 结构分析
  structure: StructureAnalysis   // 结构分析
  
  // 写作技巧
  writing_techniques: string[]   // 写作技巧
  
  // 关键词
  keywords: string[]             // 核心关键词
  
  // 改进建议
  improvement_suggestions: string[] // 可改进的点

  // 创作记录ID
  creation_id?: number           // 创作记录ID
}

// 爆款模仿请求
export interface ImitateRequest {
  reference_content: string      // 参考的爆款内容
  reference_title?: string       // 参考文章标题
  new_topic: string              // 新的主题/话题
  platform?: string              // 目标平台
  style_strength?: number        // 风格模仿强度 (0-100)
  keep_structure?: boolean       // 是否保持相同结构
  additional_requirements?: string // 额外要求
}

// 爆款模仿响应
export interface ImitateResponse {
  title: string                  // 生成的标题
  content: string                // 生成的内容
  
  // 模仿说明
  imitation_notes: string[]      // 模仿说明
  elements_applied: string[]     // 应用的爆款元素
  
  // 统计信息
  word_count: number             // 字数
  estimated_viral_score: number  // 预估爆款指数
  
  // 创作记录ID
  creation_id?: number           // 创作记录ID
}

// 爆款元素信息
export interface ViralElementInfo {
  key: string                    // 元素键名
  name: string                   // 元素名称
  description: string            // 元素描述
}

/**
 * 获取爆款元素列表
 */
export function getViralElements() {
  return request.get<{ elements: ViralElementInfo[] }>('/v1/viral/elements')
}

/**
 * 分析爆款内容
 * @param data 分析请求
 */
export function analyzeContent(data: AnalyzeRequest) {
  return request.post<AnalyzeResponse>('/v1/viral/analyze', data)
}

/**
 * 模仿爆款生成内容
 * @param data 模仿请求
 */
export function imitateContent(data: ImitateRequest) {
  return request.post<ImitateResponse>('/v1/viral/imitate', data)
}

// 内容类别对应的中文名称
export const CATEGORY_LABELS: Record<ContentCategory, string> = {
  [ContentCategory.EMOTION]: '情感共鸣',
  [ContentCategory.KNOWLEDGE]: '知识干货',
  [ContentCategory.STORY]: '故事叙事',
  [ContentCategory.NEWS]: '新闻热点',
  [ContentCategory.TUTORIAL]: '教程攻略',
  [ContentCategory.REVIEW]: '测评种草',
  [ContentCategory.OPINION]: '观点输出',
}

// 内容类别对应的颜色
export const CATEGORY_COLORS: Record<ContentCategory, string> = {
  [ContentCategory.EMOTION]: '#f759ab',
  [ContentCategory.KNOWLEDGE]: '#1890ff',
  [ContentCategory.STORY]: '#722ed1',
  [ContentCategory.NEWS]: '#fa541c',
  [ContentCategory.TUTORIAL]: '#13c2c2',
  [ContentCategory.REVIEW]: '#52c41a',
  [ContentCategory.OPINION]: '#faad14',
}

// 爆款指数等级颜色
export function getViralScoreColor(score: number): string {
  if (score >= 90) return '#52c41a'  // 绿色 - 极佳
  if (score >= 70) return '#1890ff'  // 蓝色 - 优秀
  if (score >= 50) return '#faad14'  // 橙色 - 合格
  return '#f5222d'                    // 红色 - 需提升
}

// 爆款指数等级标签
export function getViralScoreLabel(score: number): string {
  if (score >= 90) return '爆款潜力'
  if (score >= 70) return '优质内容'
  if (score >= 50) return '合格内容'
  return '需要提升'
}
